Simplify macOS build instruction

TachikakaMin opened the pull request:
cc @svekars @carljparker

TachikakaMin committed:
Update README.md

One cmakefile has different configurations for AppleClang and Clang.

TachikakaMin committed:
Update README.md

TachikakaMin commented:
@msaroufim @soumith

Hi all, 

Maybe you can check this PR since you all have changed this README MacOS compile part (:

malfet commented on README.md:
````diff
@@ -249,11 +249,20 @@ This is caused by `ld` from the Conda environment shadowing the system `ld`. You
 
 **On macOS**
 
+You should use AppleClang rather than Clang here. 
+
 ```bash
 export CMAKE_PREFIX_PATH=${CONDA_PREFIX:-"$(dirname $(which conda))/../"}
 MACOSX_DEPLOYMENT_TARGET=10.9 CC=clang CXX=clang++ python setup.py develop
 ```
 
+For default macOS configuration, you can just use 
+
+```bash
+export CMAKE_PREFIX_PATH=${CONDA_PREFIX:-"$(dirname $(which conda))/../"}
+MACOSX_DEPLOYMENT_TARGET=10.9 CC=gcc CXX=g++ python setup.py develop
+```
+
````
This is how I usually build PyTorch on MacOS for the last 3 years. I.e. if one builds using conda environment, `CMAKE_PREFIX_PATH` should already be defined. If not, no reason to tweak the variables. But I think it's important to call out that `python3` should be used, as on Intel Mac `python` still points to default Python 2.7 runtime, which is not supported for the last 3 releases of PyTorch
```suggestion
```bash
python3 setup.py develop
```

For default macOS configuration, you can just use 

```bash
export CMAKE_PREFIX_PATH=${CONDA_PREFIX:-"$(dirname $(which conda))/../"}
MACOSX_DEPLOYMENT_TARGET=10.9 CC=gcc CXX=g++ python setup.py develop
```

```

TachikakaMin replied:
You are right. 

Actually, the bug is in the file I'm referencing. AppleClang and Clang don't have --whole-archive option.

malfet replied:
`--whole-archive` is a linker option(notice, that flag is passed as `-Wl,--whole-archive`), and `-force_load` is indeed available on MacOS platform (below is the quote from `man ld`):
>      -force_load path_to_archive
>             Loads all members of the specified static archive library.  Note:
>             -all_load forces all members of all archives to be loaded.  This
>             option allows you to target a specific archive.

If you want to fix this, please do not hesitate to open a new PR and change condition from `if(CMAKE_CXX_COMPILER_ID to `if(APPLE)` or something like that

TachikakaMin replied:
That's true. I will open a new PR.

TachikakaMin replied:
New PR is here: https://github.com/pytorch/pytorch/pull/91736

malfet reviewed (changes requested):
Thank you for your interest in improving PyTorch documentation.

The file you are referencing is only relevant if one tests PyTorch for iOS/Android devices, while for a vanilla MacOS build it shouldn't make much difference (Also, if user builds  `clang` from source instead of using one provided by XCode/Developer tools, they should be pretty comfortable about the distinction between the two.)

Do you mind adjusting your change to include a very simple instruction that works for 99% of the users (i.e. when one builds locally, by default there are no need specifying deployment targets nor CC/CXX one supposed to use)

TachikakaMin committed:
Update README.md

Co-authored-by: Nikita Shulga <nikita.shulga@gmail.com>

TachikakaMin committed:
Update CMakeLists.txt

clang doesn't support --whole-archive or --no-whole-archive. We still need to use -force_load. Or there will be a compile error.

TachikakaMin commented:
The bug is in the file I'm referencing. AppleClang and Clang don't have --whole-archive option.

When I first compile code on my M1 computer, I thought it because clang(or AppleClang) in macos doesn't support  "--whole-archive" so I want to change the readme file.

malfet reviewed (changes requested):
Please do not conflate two changes into one: change `README.md` in this PR, but then update the condition in another one.

Also please note, that  current change as it stands is wrong, since if this code is executed on linux will fail, where linker indeed supports `--whole-archive` rather than `-force_load`.

TachikakaMin commented:
> Please do not conflate two changes into one: change `README.md` in this PR, but then update the condition in another one.
> 
> Also please note, that current change as it stands is wrong, since if this code is executed on linux will fail, where linker indeed supports `--whole-archive` rather than `-force_load`.

Yes, you are right.

TachikakaMin committed:
Revert "Update CMakeLists.txt"

This reverts commit 66d29026b796de470ffdec90e6c1f002cf550c74.

malfet commented on README.md:
````diff
@@ -250,10 +250,7 @@ This is caused by `ld` from the Conda environment shadowing the system `ld`. You
-MACOSX_DEPLOYMENT_TARGET=10.9 CC=clang CXX=clang++ python setup.py develop
-```
-
+python3 setup.py develop
````
```suggestion
python3 setup.py develop
```
```

malfet reviewed (changes requested):
Almost, LGTM but you forgot the closing backticks block, aren't you (and I can't propose this as a suggestion

malfet committed:
Fix syntax

malfet reviewed (approved):
Well, actually I can, looks good to me now :)

malfet commented:
@pytorchbot merge -f "Lint is green"
